File: java/src/Test2.java
import java.util.Scanner;
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;
import java.io.*;
import java.util.NoSuchElementException;

public class Test2 extends Test {
    public boolean runTest(File f, boolean verbose) {
	List<Integer> measurements = TestHelpers.getIntegerListFromFile(f);
	Integer expected = getResults(f, verbose);
	Integer actual = -1;

	if ((expected != null) && (expected < 0)) {
	    // bad results file
	    return false;
	}

	try {
	    actual = Problem2.raindrops(measurements);
	} catch (Exception e) {
	    if (verbose) {
		System.out.println("Unexpected exception: " + e);
	    }
	    return false;
	}

	if (actual == null) {
	    if (expected == null) {
		return true;
	    } else {
		if (verbose) {
		    System.out.println("Failed: actual result is null");
		}
		return false;
	    }
	} else if (actual.equals(expected)) {
	    return true;
	} else {
	    if (verbose) {
		System.out.println("Failed actual result and expected result do not match");
		System.out.println(String.format("    actual is %d", actual));
		System.out.println(String.format("    expected is %d", expected));
	    }
	    return false;
	}
    }

    private static Integer getResults(File inputFile, boolean verbose) {
	Integer result = -1;
	Scanner scanner = TestHelpers.getResultsFileScanner(inputFile);

	if (scanner == null) {
	    return result;
	}

	try {
	    String data = scanner.nextLine();
	    if (data.equals("INSUFFICIENT DATA")) {
		result = null;
	    } else {
		result = Integer.parseInt(data);
	    }
	} catch (NumberFormatException e) {
	    if (verbose) {
		System.out.println("Results file has the wrong format");
	    }
	}

	return result;
    }

    public static void main(String[] args) {
	System.exit(TestHelpers.doTests(args, new Test2()));
    }
}

File: cpp/src/Test2.cpp
#include "TestHelpers.h"
#include "Problem2.h"
#include <cstdio>
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <optional>

class Test {
public:
static std::optional<int> getResult(std::string& f) {
     std::string resultsFileName = TestHelpers::getResultsFile(f);
     std::ifstream ifs(resultsFileName);
     std::optional<int> res;

     std::string data;
     if (ifs) {
         ifs >> data;
         if (std::isdigit(data[0])) {
             res = std::stoi(data);
         } else {
             res = std::nullopt;
         }
     } else {
         std::cout << "Results file not found: " << resultsFileName << std::endl;;
         exit(1);
     }
     return res;
 }

static bool runTest(std::string f) {
    std::vector<int> measurements = TestHelpers::getData(f);
    std::optional<int> actual = Problem2::raindrops(measurements);
    std::optional<int> expected = getResult(f);

    if (actual == std::nullopt) {
        if (expected == std::nullopt) {
            return true;
        } else {
            std::cout << "Failed: actual result is null\n";
	    return false;
        }
    } else if (actual == expected) {
        return true;
    } else {
        std::cout << "Failed actual result and expected result do not match\n";
        std::string actualVal = actual ? std::to_string(actual.value()) : "null";
        std::string expectedVal = expected ? std::to_string(expected.value()) : "null";
        std::cout << "    actual is " << actualVal << std::endl;
        std::cout << "    expected is " << expectedVal << std::endl;
	return false;
    }
    
}

  


};



int main (int argc, char* argv[]) {
    std::vector<std::string> files = TestHelpers::getFiles(argc, argv);
    if (files.size() == 0) {
        std::cout << "no test file found"; 
        exit(1);
    } 

    int count = 0;
    for (size_t i = 0; i < files.size(); ++i) {
        std::string f = files[i];
	std::cout << "Test " << f << ": ";	 	
        if (Test::runTest(f)) {
	  std::cout << "PASSED\n";	  
	  count++;
	} else {
	  std::cout << "FAILED\n";	  
	}
    }

    std::cout << "Files: " << files.size();
    std::cout << "count: " << count;
    if (count == files.size()) {
      return 42;
    } else {
      return 0;
    }
}
